imperial_coldfront_plugin/dart.py

```diff
@@ -0,0 +1,37 @@
+"""Module for creation and validation of Dart ID AllocationAttribute's."""
+
+from coldfront.core.allocation.models import (
+    Allocation,
+    AllocationAttribute,
+    AllocationAttributeType,
+)
+
+
+class DartIDValidationError(Exception):
+    """Error when a Dart ID does not meet validation criteria."""
+
+
+def validate_dart_id(dart_id: str):
+    """Validate range and availability of Dart ID value."""
+    try:
+        if int(dart_id) < 1:
+            raise DartIDValidationError("Dart ID outside valid range")
+    except ValueError:
+        raise DartIDValidationError("Dart ID is not a number")
+    if AllocationAttribute.objects.filter(
+        allocation_attribute_type__name="DART ID", value=dart_id
+    ):
+        raise DartIDValidationError("Dart ID already assigned to an allocation")
+
+
+def create_dart_id_attribute(dart_id: str, allocation: Allocation):
+    """Create an AllocationAttribute for a Dart ID."""
+    validate_dart_id(dart_id)
+    dart_id_attribute_type = AllocationAttributeType.objects.get(
+        name="DART ID", is_changeable=False
+    )
+    return AllocationAttribute.objects.create(
+        allocation_attribute_type=dart_id_attribute_type,
+        allocation=allocation,
+        value=dart_id,
+    )
```

Comment:
I understand that this is just the way Coldfront do things, but do you know what's the rationale for this approach rather than just having a `DartID` model with `Allocation` as foreign key and a value? I'm sure there's a good reason for this - I would say that facilitating having arbitrary number of attributes without an explosion of models for each of them - but just checking to make sure I understand this right.

Reply:
So Coldfront has implemented a fairly complex data model that allows for attaching arbitrary metadata to allocations via these atributes. I think the rationale is supposed to be that you can create attribute types etc. all via the admin interface without the need to write code. I guess that's kind of ok if you just want to manually tag arbitrary metadata but it's not a great approach IMO if you want to build more sophisticated behaviour around those attributes as you just end up having to handle these things via data migrations and it's all rather cumbersome.